internal/metrics/v2/histogram_test.go

```diff
@@ -0,0 +1,176 @@
+					LabelNames: []string{"histogram_test_label"},
+				},
+			},
+		},
+		{
+			name: "should panic when HistogramOpts is empty",
+			args: args{
+				registry: prometheus.NewRegistry(),
+				opts:     HistogramOpts{},
+			},
+			shouldPanic: true,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if test.shouldPanic {
+				require.Panics(t, func() {
+					NewHistogram(test.args.registry, test.args.opts)
+				})
+			} else {
+				require.IsType(t, &prometheusHistogram{}, NewHistogram(test.args.registry, test.args.opts))
+			}
+		})
+	}
+}
+
+func Test_prometheusHistogram_Observe(t *testing.T) {
+	type fields struct {
+		registry *prometheus.Registry
+		opts     HistogramOpts
+	}
+	type args struct {
+		floatVals []float64
+		label     []Label
+	}
+	type expected struct {
+		sampleSum   float64
+		sampleCount uint64
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+		expect expected
+	}{
+		{
+			name: "should observe",
+			fields: fields{
+				registry: prometheus.NewRegistry(),
+				opts: HistogramOpts{
+					Subsystem:  "cp",
+					Name:       "histogram_test_total",
+					Help:       "histogram_test help",
+					Buckets:    []float64{1, 2, 3, 4, 5, 6},
+					LabelNames: []string{},
+				},
+			},
+			args: args{
+				floatVals: []float64{1, 1, 2, 2},
+				label:     []Label{},
+			},
+			expect: expected{
+				sampleSum:   float64(6),
+				sampleCount: uint64(4),
+			},
+		},
+		{
+			name: "should observe with labels",
+			fields: fields{
+				registry: prometheus.NewRegistry(),
+				opts: HistogramOpts{
+					Subsystem:  "cp",
+					Name:       "histogram_test_total",
+					Help:       "histogram_test help",
+					Buckets:    []float64{1, 2, 3, 4, 5, 6},
+					LabelNames: []string{"foo", "bar"},
+				},
+			},
+			args: args{
+				floatVals: []float64{1, 1, 2, 2},
+				label:     []Label{{"foo", "fooval"}, {"bar", "barval"}},
+			},
+			expect: expected{
+				sampleSum:   float64(6),
+				sampleCount: uint64(4),
+			},
+		},
+		{
+			name: "should observe when input is < 0",
+			fields: fields{
+				registry: prometheus.NewRegistry(),
+				opts: HistogramOpts{
+					Subsystem:  "cp",
+					Name:       "histogram_test_total",
+					Help:       "histogram_test help",
+					Buckets:    []float64{1, 2, 3, 4, 5, 6},
+					LabelNames: []string{"foo", "bar"},
+				},
+			},
+			args: args{
+				floatVals: []float64{-1, -1, -2, -2},
+				label:     []Label{{"foo", "fooval"}, {"bar", "barval"}},
+			},
+			expect: expected{
+				sampleSum:   float64(-6),
+				sampleCount: uint64(4),
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			h := NewHistogram(test.fields.registry, test.fields.opts)
+			for _, val := range test.args.floatVals {
+				h.Observe(val, test.args.label...)
+			}
+			family, err := test.fields.registry.Gather()
+			require.NoError(t, err)
+			histogram := family[0].Metric[0].Histogram
```

Comment:
As a sanity check, we should assert the length of `family` & `family[0].Metric` (and in all the other places).

Reply:
[152f5a8](https://github.com/Kong/koko/pull/364/commits/152f5a8e502fc416b30574a645dd7f618a7d5887)